tajo-core/src/main/java/org/apache/tajo/querymaster/DefaultTaskScheduler.java

```diff
@@ -251,63 +250,70 @@ public void handle(TaskSchedulerEvent event) {
     }
   }
 
-  @Override
-  public void handleTaskRequestEvent(TaskRequestEvent event) {
-
-    taskRequests.handle(event);
-    int hosts = scheduledRequests.leafTaskHostMapping.size();
+  private Set<Integer> getWorkerIds(Collection<String> hosts){
+    Set<Integer> workerIds = Sets.newHashSet();
+    if(hosts.isEmpty()) return workerIds;
 
-    // if available cluster resource are large then tasks, the scheduler thread are working immediately.
-    if(remainingScheduledObjectNum() > 0 &&
-        (remainingScheduledObjectNum() <= hosts || hosts <= taskRequests.size())){
-      synchronized (schedulingThread){
-        schedulingThread.notifyAll();
+    for (WorkerConnectionInfo worker : stage.getContext().getWorkerMap().values()) {
+      if(hosts.contains(worker.getHost())){
+        workerIds.add(worker.getId());
       }
     }
+    return workerIds;
   }
 
-  @Override
-  public int remainingScheduledObjectNum() {
-    return scheduledObjectNum;
-  }
 
-  private class TaskRequests implements EventHandler<TaskRequestEvent> {
-    private final LinkedBlockingQueue<TaskRequestEvent> taskRequestQueue =
-        new LinkedBlockingQueue<TaskRequestEvent>();
+  protected LinkedList<TaskRequestEvent> createTaskRequest() throws Exception {
+    LinkedList<TaskRequestEvent> taskRequestEvents = new LinkedList<TaskRequestEvent>();
 
-    @Override
-    public void handle(TaskRequestEvent event) {
-      if(LOG.isDebugEnabled()){
-        LOG.debug("TaskRequest: " + event.getContainerId() + "," + event.getExecutionBlockId());
-      }
+    int requestContainerNum = Math.max(remainingScheduledObjectNum(), 1);
```

Comment:
Would you give a explanation on the above line? 
Even though ```remainingScheduledObjectNum()``` is 0, requestContainerNum is 1. I wonder why at least one container is requested even though there is no remaining scheduled objects.